File: src/classes/Explosion.java
//****************************************************************************
// ----- general information -----
//
// Explosion.java -- Ship Explosion
//
// Written by:				Neal Lawson, e-mail: nlawson@uga.icad.edu
// Initial Release:		01/29/97.
//
// Copyright (c) Neal Lawson, 1997
//
// ----- version information -----
// v 1.00a,	12/20/96 - 01/29/97,	Initial Classes and testing.
//
// ----- history and repairs -----
//
// ----- Description -----
// Explosion is a Mover that is an animated ship explosion.
//****************************************************************************
/*---------------------------------------------------------------------------
	Many of the concepts of this class and game were generously taken from
	Chris Boyke's game, Spacewar.  Here is Chris' copyright statement and
	address:

   Copyright (C) 1996 by Chris Boyke, Clear View Consulting
                http://www.clear-view.com
 *--------------------------------------------------------------------------*/

import Mover;
import java.awt.*;
import Gameutil;

/*
 *
 *	Explosion
 *
 */
class Explosion extends Mover {
	static final int MAX_SIZE = 40;
	static final int SIZE_INC = 4;
	static final double SPEED_MUL = 0.95;
	int e_size;			// explosion size
	Point wp, vp;		// world point -> view point

	public Explosion(int x, int y, int sx, int sy)	
	{
		e_size = 0;
		m_x = x;
		m_y = y;
		m_xvel = sx;
		m_yvel = sy;
		m_alive = true;
		wp = new Point(m_x, m_y);
		vp = new Point(m_x, m_y);
	}

	public void tick() {
		// Slow down
		m_xvel *= SPEED_MUL;
		m_yvel *= SPEED_MUL;

		// Grow the explosion
		e_size += SIZE_INC;
		if (e_size >= MAX_SIZE)
			m_alive = false;

		// Move
		m_x += m_xvel;
		m_y += m_yvel;
		super.tick();
	}

	public void paint(Graphics g)
	{
		wp.x = m_x;
		wp.y = m_y;

		parent.gc_vp.wp.Worldpoint2Viewpoint(parent.gc_vp, wp, vp);
		g.setColor(Gameutil.randomColor());
//		g.setColor(Color.white);
		g.fillOval(vp.x-e_size, vp.y-e_size, e_size*2, e_size*2);
	}
}

File: src/classes/Explosion.ts
// import { Graphics, Point } from "../../node_modules/pixi.js/dist/pixi.mjs";
import { Graphics, Point } from "pixi.js";
import { Mover } from "./Mover.js";
import { GameConstants } from "./GameConstants.js";
import { GameUtils } from "./GameUtils.js";
import { Viewport } from "./Engine2D.js";


//****************************************************************************
// ----- general information -----
//
// Explosion.java -- Ship Explosion
//
// Written by:				Neal Lawson, e-mail: nlawson@uga.icad.edu
// Initbreak
// ----- history and repairs -----
//
// ----- Description -----
// Explosion is a Mover that is an animated ship explosion.
//****************************************************************************
                
export class Explosion extends Mover {
    static MAX_SIZE = [30, 60, 90, 300];
    static SIZE_INC = [1, 3, 5, 8];
    static COLOR = [0xff0000, 0xF5B507, 0xF7E96A, 0xffffff];
    static SPEED_MUL = 0.95;
    private magnitude: number;      // "Class" of explosion - use as index in above arrays.
    private cur_size = 0;		// explosion size at the moment.
    private vp: Viewport;
    // private world_point: Point;      // world point -> view point
    // private view_point: Point;		

    constructor(vp: Viewport, x: number, y: number, xv: number, yv: number, magnitude: number)
    {
        super(vp.wp, Mover.TOPO_WRAP, x, y, xv, yv);

        // magnitude is used as an offset into an array. Bad form. Bounds check it.
        if (magnitude < 0) magnitude = 0;
        if (magnitude > 3) magnitude = 3;

        this.magnitude = magnitude;
        this.vp = vp;
    }

    tick(): void
    {
        // Slow down
        this.xvel *= Explosion.SPEED_MUL;
        this.yvel *= Explosion.SPEED_MUL;

        // Grow the explosion
        this.cur_size += Explosion.SIZE_INC[this.magnitude];

        // Explosion is done?
        if (this.cur_size >= Explosion.MAX_SIZE[this.magnitude])
            this.die();

        // Move
        this.x += this.xvel;
        this.y += this.yvel;
        super.tick();
    }

    paint(g: Graphics): void
    {
        const world_point = new Point(this.x, this.y);
        const view_point = new Point(this.x, this.y);
        this.vp.Worldpoint2Viewpoint(world_point, view_point);

   		g.lineStyle(2, 0xdd0000, 1);
        g.beginFill(Explosion.COLOR[this.magnitude], 0.8);
		g.drawCircle(view_point.x, view_point.y, this.cur_size*2);
        g.endFill();
//         g.setColor(Gameutil.randomColor());
// //		g.setColor(Color.white);
//         g.fillOval(view_point.x-size, view_point.y-size, size*2, size*2);
    }
}